compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2252,6 +2254,13 @@ impl<'a> Parser<'a> {
     /// Parses a `let $pat = $expr` pseudo-expression.
     /// The `let` token has already been eaten.
     fn parse_let_expr(&mut self, attrs: AttrVec) -> PResult<'a, P<Expr>> {
+        if !self.is_in_a_valid_let_ctx {
+            self.struct_span_err(
+                self.prev_token.span,
+                "expected expression, found `let` statement",
+            )
+            .emit();
+        }
```

Comment:
Don't know why but returning `Err` here invalidates any following invalid `let`;

```rust
    &let 0 = 0; // Error!

    !let 0 = 0; // Ignored!
```

Reply:
It's because of how you're enabling and disabling `is_in_a_valid_let_ctx`. I'll leave a comment in-line.

Reply:
Already tried
```rust
self.is_in_a_valid_let_ctx = true;
let cond = match self.parse_expr_res(Restrictions::NO_STRUCT_LITERAL, None) {
  Err(err) => {
    self.is_in_a_valid_let_ctx = false;
    return Err(err);
  }
  Ok(elem) => {
    self.is_in_a_valid_let_ctx = false;
    elem
  }
}
```

But the end result was the same 🤷